File: cloud-reactor/src/main/java/com/sequenceiq/cloudbreak/cloud/scheduler/SyncPollingScheduler.java
package com.sequenceiq.cloudbreak.cloud.scheduler;

import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.concurrent.atomic.AtomicInteger;

import org.springframework.stereotype.Component;

import com.dyngr.Polling;
import com.dyngr.core.AttemptResults;
import com.dyngr.exception.PollerStoppedException;
import com.dyngr.exception.UserBreakException;
import com.sequenceiq.cloudbreak.cloud.task.PollTask;

@Component
public class SyncPollingScheduler<T> {

    private static final int POLLING_INTERVAL = 1000;

    private static final int MAX_POLLING_ATTEMPT = 5000;

    private static final int FAILURE_TOLERANT_ATTEMPT = 3;

    public T schedule(PollTask<T> task) throws Exception {
        return schedule(task, POLLING_INTERVAL, MAX_POLLING_ATTEMPT, FAILURE_TOLERANT_ATTEMPT);
    }

    public T schedule(PollTask<T> task, int interval, int maxAttempt, int maxFailureTolerant) throws Exception {
        AtomicInteger actualFailureTolerant = new AtomicInteger(0);

        try {
            return Polling.stopAfterAttempt(maxAttempt)
                    .waitPeriodly(interval, TimeUnit.MILLISECONDS)
                    .stopIfException(true)
                    .run(() -> {
                        if (task.cancelled()) {
                            return AttemptResults.breakFor(new CancellationException("Task was cancelled."));
                        }
                        try {
                            T callResult = task.call();
                            if (task.completed(callResult)) {
                                return AttemptResults.finishWith(callResult);
                            }
                        } catch (Exception ex) {
                            int currentTolerant = actualFailureTolerant.incrementAndGet();
                            if (currentTolerant >= maxFailureTolerant) {
                                return AttemptResults.breakFor(ex);
                            }
                        }
                        return AttemptResults.justContinue();
                    });
        } catch (PollerStoppedException e) {
            throw new TimeoutException(String.format("Task (%s) did not finish within %d seconds", task.getClass().getSimpleName(), interval * maxAttempt));
        } catch (UserBreakException e) {
            throw (Exception) e.getCause();
        }
    }
}

File: cloud-reactor/src/main/java/com/sequenceiq/cloudbreak/cloud/scheduler/SyncPollingScheduler.kt
package com.sequenceiq.cloudbreak.cloud.scheduler

import java.util.concurrent.ExecutionException
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException

import javax.inject.Inject

import org.springframework.stereotype.Component

import com.google.common.util.concurrent.ListenableScheduledFuture
import com.google.common.util.concurrent.ListeningScheduledExecutorService
import com.sequenceiq.cloudbreak.cloud.task.FetchTask
import com.sequenceiq.cloudbreak.cloud.task.PollTask

@Component
class SyncPollingScheduler<T> {

    @Inject
    private val scheduler: ListeningScheduledExecutorService? = null

    @Throws(ExecutionException::class, InterruptedException::class, TimeoutException::class)
    @JvmOverloads fun schedule(task: PollTask<T>, interval: Int = POLLING_INTERVAL, maxAttempt: Int = MAX_POLLING_ATTEMPT, maxFailureTolerant: Int = FAILURE_TOLERANT_ATTEMPT): T {
        var result: T? = null
        var actualFailureTolerant = 0
        for (i in 0..maxAttempt - 1) {
            if (task.cancelled()) {
                throw CancellationException("Task was cancelled.")
            }
            try {
                val ft = schedule(task, interval)
                result = ft.get()
                if (task.completed(result)) {
                    return result
                }
            } catch (ex: Exception) {
                actualFailureTolerant++
                if (actualFailureTolerant >= maxFailureTolerant) {
                    throw ex
                }
            }

        }
        throw TimeoutException(String.format("Task did not finished within %d seconds", interval * maxAttempt))
    }

    fun schedule(task: FetchTask<T>, interval: Int): ListenableScheduledFuture<T> {
        return scheduler!!.schedule(task, interval.toLong(), TimeUnit.SECONDS)
    }

    companion object {

        private val POLLING_INTERVAL = 5
        private val MAX_POLLING_ATTEMPT = 1000
        private val FAILURE_TOLERANT_ATTEMPT = 3
    }

}
